src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp

```diff
@@ -10,6 +10,270 @@
+      for (size_t c = 0; c < SpatialDim + 1; ++c) {
+        for (size_t d = 0; d < SpatialDim + 1; ++d) {
+          constraint->get(i, a) -= 0.5 * spacetime_normal_one_form.get(a) *
+                                   inverse_spacetime_metric.get(c, d) *
+                                   d_pi.get(i, c, d);
+        }
+      }
+    }
+  }
+}
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+void two_index_constraint_add_term_5_of_11(
+    gsl::not_null<tnsr::ia<DataType, SpatialDim, Frame>*> constraint,
+    const tnsr::ia<DataType, SpatialDim, Frame>& d_gauge_function) noexcept {
+  for (size_t i = 0; i < SpatialDim; ++i) {
+    for (size_t a = 0; a < SpatialDim + 1; ++a) {
+      constraint->get(i, a) += d_gauge_function.get(i, a);
+    }
+  }
+}
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+void two_index_constraint_add_term_6_of_11(
+    gsl::not_null<tnsr::ia<DataType, SpatialDim, Frame>*> constraint,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_normal_vector,
+    const tnsr::a<DataType, SpatialDim, Frame>& spacetime_normal_one_form,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi,
+    const tnsr::AA<DataType, SpatialDim, Frame>&
+        inverse_spacetime_metric) noexcept {
+  for (size_t i = 0; i < SpatialDim; ++i) {
+    for (size_t a = 0; a < SpatialDim + 1; ++a) {
+      for (size_t c = 0; c < SpatialDim + 1; ++c) {
+        for (size_t d = 0; d < SpatialDim + 1; ++d) {
+          for (size_t e = 0; e < SpatialDim + 1; ++e) {
+            for (size_t f = 0; f < SpatialDim + 1; ++f) {
+              if (a > 0) {
+                constraint->get(i, a) += 0.5 * phi.get(a - 1, c, d) *
+                                         phi.get(i, e, f) *
+                                         inverse_spacetime_metric.get(c, e) *
+                                         inverse_spacetime_metric.get(d, f);
+              }
+              for (size_t j = 0; j < SpatialDim; ++j) {
+                constraint->get(i, a) +=
+                    0.5 * spacetime_normal_vector.get(j + 1) *
+                    spacetime_normal_one_form.get(a) * phi.get(j, c, d) *
+                    phi.get(i, e, f) * inverse_spacetime_metric.get(c, e) *
+                    inverse_spacetime_metric.get(d, f);
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+void two_index_constraint_add_term_7_of_11(
+    gsl::not_null<tnsr::ia<DataType, SpatialDim, Frame>*> constraint,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi,
+    const tnsr::AA<DataType, SpatialDim, Frame>& inverse_spacetime_metric,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_normal_vector,
+    const tnsr::a<DataType, SpatialDim, Frame>&
+        spacetime_normal_one_form) noexcept {
+  for (size_t i = 0; i < SpatialDim; ++i) {
```

Comment:
I went the route of minimizing allocations and of maximizing code clarity, as these terms are so  complicated. I’d feel better optimizing after the terms are reviewed for correctness in their current form, preferably in a future PR, but I’m happy to benchmark with perf

Reply:
My guess is that if a Variables is used then it will be more efficient to precompute some things, but of course that's "pure speculation" (it should be measured :) ). Having said that, I would suggest going with the old "1 good measurement is worth thousands of expert opinions" and measure. Though I'm also fine focusing on correctness in this PR and doing a follow-up PR that explores performance. I wouldn't be all that surprised if the loop structure and whether or not a temporary buffer is worth the allocation cost will depend on the number of grid points (small number of grid points allocations might be too expensive). Hope that helps! :)

Reply:
I, too, am happy to keep performance changes for a later PR...

Reply:
Great. I’ll start playing with perf, so once this PR is in, I’ll be in a good place to profile and optimize as needed

Reply:
If you're interested in what I did for definite integral I wrote up a bit on it here: https://github.com/nilsdeppe/spectre/wiki/Benchmark-Codes It shows how you could run over a series of different mesh sizes, for example. I'd recommend running perf on the Benchmark executable.

Reply:
Thanks! I’ll take a look, but I’ll focus on fixes to this PR first, then profiling and optimization for the next one.